apps/dapp-console-api/src/routes/contracts/ContractsRoute.ts

```diff
@@ -48,6 +52,95 @@ export class ContractsRoute extends Route {
       }
     })
 
+  public readonly createContract = 'createContract' as const
+  public readonly createContractController = this.trpc.procedure
+    .use(isPrivyAuthed(this.trpc))
+    .input(
+      this.z.object({
+        contractAddress: zodEthereumAddress,
+        deploymentTxHash: zodEthereumTransaction,
+        deployerAddress: zodEthereumAddress,
+        chainId: this.z.number(),
+        appId: this.z.string(),
+      }),
+    )
+    .mutation(async ({ ctx, input }) => {
+      const { user } = ctx.session
+      const deploymentTxHash = input.deploymentTxHash as Hash
+      const contractAddress = input.contractAddress as Address
+      const deployerAddress = input.deployerAddress as Address
+      const { chainId, appId } = input
+
+      if (!user) {
+        throw Trpc.handleStatus(401, 'user not authenticated')
+      }
+
+      const publicClient = supportedChainsPublicClientsMap[chainId]
+
+      if (!publicClient) {
+        throw Trpc.handleStatus(400, 'chain not supported')
+      }
+
+      const deploymentTx = await publicClient
+        .getTransactionReceipt({
+          hash: deploymentTxHash,
+        })
+        .catch((err) => {
+          metrics.listContractsErrorCount.inc({ chainId })
+          this.logger?.error(
+            {
+              error: err,
+              entityId: ctx.session.user?.entityId,
+              txHash: deploymentTxHash,
+              chainId,
+            },
+            'error fetching tx',
+          )
+          throw Trpc.handleStatus(500, 'error fetching deployment transaction')
+        })
+
+      if (deploymentTx.from !== deployerAddress) {
+        throw Trpc.handleStatus(
+          400,
+          'deployer address does not match deployment transaction',
+        )
+      }
+
+      if (deploymentTx.contractAddress !== contractAddress) {
+        throw Trpc.handleStatus(
+          400,
+          'contract was not created by deployment transaction',
+        )
+      }
+
```

Comment:
just a heads up, might be a feature we add on later - this doesn't support create2. we will likely need a more robust way to track deployed contract addresses

example: https://sepolia-optimism.etherscan.io/tx/0xadad1253cd554bbb835b03ae4f5aa9b4534e87379cfb2ebe9f3f297853df3a25

```typescript
{
  blockHash:
    '0x7a88caf2070772c49672acbb90452210ad6df135453af8ebcc5f87ab573bbc50',
  blockNumber: '8676011n',
  contractAddress: null,
  cumulativeGasUsed: '1109129n',
  effectiveGasPrice: '3000000252n',
  from: '0x8a986ce389686dd140a5fd834b8b25d9d053d0fe',
  gasUsed: '1065266n',
  l1Fee: '107132928848475n',
  l1GasPrice: '79352n',
  l1GasUsed: '75940n',
  logs: [],
  logsBloom:
    '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  status: 'success',
  to: '0x4e59b44847b379578588920ca78fbf26c0b4956c',
  transactionHash:
    '0xadad1253cd554bbb835b03ae4f5aa9b4534e87379cfb2ebe9f3f297853df3a25',
  transactionIndex: 1,
  type: 'eip1559',
  l1FeeScalar: null,
}

```

https://sepolia-optimism.etherscan.io/tx/0xadad1253cd554bbb835b03ae4f5aa9b4534e87379cfb2ebe9f3f297853df3a25

Reply:
I believe for a create2 deploy the `to` address is enough to let us know what the contract address is. An initial version of this can probably check to see if there was a contract address on the tx. If not it can fall back to checking the `to` address, and that it's bytecode is not `0x`. There will be complications when it comes to them having to verify any challenges for the create2 deployer they used and we'll likely need to add logic in the future thats lets us traverse a couple of txs to figure out whos the real owner.

Reply:
I think in the above transaction, the `to` address is `0x4e59b44847b379578588920ca78fbf26c0b4956c` the create2 deployer, and the created contract address is `0xab559628b94fd9748658c46e58a85efb52fdaca6`. We might have to track the internal calls (https://sepolia-optimism.etherscan.io/tx/0xadad1253cd554bbb835b03ae4f5aa9b4534e87379cfb2ebe9f3f297853df3a25#internal)  to get the final address

Reply:
yea thats true, don't worry about this flow @tremarkley we'll deal with it later

Reply:
added an issue for this: https://github.com/ethereum-optimism/ecopod/issues/929